### PR TITLE
Feature/protected routes

### PR DESCRIPTION
**Pull request by choonkending**
# Related Issues
#47 #50 #51

Protected routes example. I'm still thinking what's the best way to render user specific data.

Also note: `willTransitionTo` needs to be changed once react-router 1.x is stable.

### CODE REVIEW

**Comment by choonkending** [comment]
Thinking of a better solution: which is to grab the state from store, but `willTransitionTo` won't work in this case. 

https://github.com/rackt/react-router/pull/590#issuecomment-111910388 seems like a very convincing argument to update bump react-router to `1.x`.

**Comment by joshhornby** [comment]
I don't see a reason not to bump up to `1.x` - Seems a sensible approach.

I think you mentioned it above but once I redirect a user to the page they initially requested I don't have the full user object just the state which has been set by the login. But I think you know about this!

**Comment by choonkending** [comment]
@joshhornby Yep. I'll be working on this tonight and hopefully can get it done for you to test.

**Comment by choonkending** [comment]
Might still get an error saying: 

```Failed propType: Invalid prop `location` supplied to `Router`, expected instance of `Location```

But it won't cause any problems (that I can find so far)
https://github.com/rackt/react-router/issues/1524

**Comment by choonkending** [comment]
One last checkbox:
- [ ] Add the auth to UserStore object, so we can grab it from client and server. Then this PR will be complete. 

Yay. Sleep time.

**Comment by choonkending** [comment]
@joshhornby I've done it. You can have a play with it now. 

If you want to access the UserStore object, your Dashboard component needs to listen to the Store. (In the future I'll be using AltContainer for the entire App).

File: app/utils/UserWebAPIUtils.js [hunk 1]
```diff
@@ -35,6 +35,13 @@ const utils = {
       contentType: 'application/json',
       data: JSON.stringify(data)
     });
+  },
```
Comment (joshhornby) [inline app/utils/UserWebAPIUtils.js]: This can be deleted.

File: server/config/routes.js [hunk 2]
```diff
@@ -15,6 +15,7 @@ module.exports = function(app, passport) {
   app.post('/login', users.postLogin);
   app.post('/signup', users.postSignUp);
   app.get('/logout', users.getLogout);
+  app.get('/loggedin', users.getLoggedIn);
```
Comment (joshhornby) [inline server/config/routes.js]: This can be deleted.

File: server/controllers/users.js [hunk 3]
```diff
@@ -56,3 +56,16 @@ exports.postSignUp = function(req, res, next) {
   });
 };
 
+/**
```
Comment (joshhornby) [inline server/controllers/users.js]: This can be deleted.

**Comment by joshhornby** [comment]
@choonkending Looks really good and seems to work for me.

Might this be a good chance to also solve this issue https://github.com/choonkending/react-webpack-node/issues/56?

Should be a case of creating another ```<Route>``` in the routes file such as 

```
<Route name="unauthorisedApp" path="/auth" component={UnauthorisedApp}>
   <Route path="logout" name="logout" component={Logout} />
   <Route path="login" name="login" component={Login} />
</Route>
```

The only issue I can't work out is the correct place to 'redirect' the user to their route. Good work though :+1:

**Comment by choonkending** [comment]
@joshhornby Thanks for that!

Hmm. I think we should probably address #56 as another PR, so this won't be a never-ending PR.

I agree with the login component, I think its current position isn't ideal.

> The only issue I can't work out is the correct place to 'redirect' the user to their route.

What do you mean? As in a Redirect route to when they hit a 404? Or when they meet a certain condition?

**Comment by joshhornby** [comment]
@choonkending Fair point and I agree.

Ignore me for now, this is an issue I am having with my set up!

**Comment by choonkending** [comment]
Alright. I will merge this into master in a couple of hours. :)

This was a good journey with react-router beta. :football:

**Pull request merged**